chore: upgrade altbeacon to 2.13.1

dopry opened the pull request:
see: https://github.com/AltBeacon/android-beacon-library/releases

2.13.1 Changelog

* fix: incorrect immediateScanJobId meta-data name in manifest causing crash

2.13 Changelog

* feat: Add BluetoothMedic to fix crashing bluetooth stacks
* feat: Allow configuring job ids
* fix: Allow scans with screen off on Android 8.1

depends on: #363

dopry committed:
fix: add reference to com.android.support.support-core-utils

* including only core-utils to streamline size, 1.23MB smaller than support-v4
   in demo:  v4 size:  12.10MB, core-utils size: 10.87Mb

dopry committed:
chore: upgrade altbeacon to 2.13.1

see: https://github.com/AltBeacon/android-beacon-library/releases

2.13.1 Changelog

* fix: incorrect immediateScanJobId meta-data name in manifest causing crash

2.13 Changelog

* feat: Add BluetoothMedic to fix crashing bluetooth stacks
* feat: Allow configuring job ids
* fix: Allow scans with screen off on Android 8.1

dopry commented:
The current PR is broken. The AAR dependency isn't supported as a lib-file. I chose to use an AAR since it was easy to download from https://github.com/AltBeacon/android-beacon-library/releases and add to the code base.   To utilize the AAR it looks like I will need to add a gradle file and declare the AAR as a  dependency and add a resource-file entry in the plugin.xml to copy the .arr  to where gradle expects the libs to live.  I'll update shortly.

dopry commented:
I've updated with gradle for AAR support.

dopry commented:
If you want to test and you're not comfortable with working with plugins locally you can try installing from my fork

```
cordova plugin add https://github.com/dopry/cordova-plugin-ibeacon#1e6e7a05b564532cc37e22f3c79388e8e94e5de0
```

that commit also includes https://github.com/petermetz/cordova-plugin-ibeacon/pull/363

jakeleveroni commented:
@dopry So im working on trying this now, and i get this error:

```
> Could not find :android-beacon-library-2.13.1:.
  Searched in the following locations:
      file:/Users/jleveroni/Library/Android/sdk/extras/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
      file:/Users/jleveroni/Library/Android/sdk/extras/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
      file:/Users/jleveroni/Library/Android/sdk/extras/google/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
      file:/Users/jleveroni/Library/Android/sdk/extras/google/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
      file:/Users/jleveroni/Library/Android/sdk/extras/android/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
      file:/Users/jleveroni/Library/Android/sdk/extras/android/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
      https://jcenter.bintray.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
      https://jcenter.bintray.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
      https://maven.google.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
      https://maven.google.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
      https://repo1.maven.org/maven2//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
      https://repo1.maven.org/maven2//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
      file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/libs/android-beacon-library-2.13.1-.aar
      file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/libs/android-beacon-library-2.13.1.aar
  Required by:
      project :app
```

The very last path it tried seems close to correct because the altbeacon .aar file is located at

file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/**src/main**/libs/android-beacon-library-2.13.1.aar

Not

file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/libs/android-beacon-library-2.13.1.aar

Im not very experienced at all with gradle, is there a way to change the path that it searches for that .aar file in? Or perhaps my ionic build is misconfigured and the libs/ folder is supposed to be directly under the app/ folder?

EDIT: 
I pulled in the changes from both your PRs (https://github.com/jakeleveroni/cordova-plugin-ibeacon/commit/fe987f245fcdb0418badf037c1329ca9c6326a95)
and changed the following:

```gradle
repositories{
    jcenter()
    flatDir{
        dirs 'libs'
    }
}
```
 To

```gradle
repositories{
    jcenter()
    flatDir{
        dirs 'src/main/libs'
    }
}
```
In the `/platforms/android/com.unarin.cordova.beacon/starter-cordova-plugin-ibeacon.gradle`

And everything works like a charm!

dopry commented:
Not home, I'll check where my build puts it tomorrow.

On Thu, Apr 19, 2018, 5:40 PM Jacob <notifications@github.com> wrote:

> @dopry <https://github.com/dopry> So im working on trying this now, and i
> get this error:
>
> > Could not find :android-beacon-library-2.13.1:.
>   Searched in the following locations:
>       file:/Users/jleveroni/Library/Android/sdk/extras/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
>       file:/Users/jleveroni/Library/Android/sdk/extras/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
>       file:/Users/jleveroni/Library/Android/sdk/extras/google/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
>       file:/Users/jleveroni/Library/Android/sdk/extras/google/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
>       file:/Users/jleveroni/Library/Android/sdk/extras/android/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
>       file:/Users/jleveroni/Library/Android/sdk/extras/android/m2repository//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
>       https://jcenter.bintray.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
>       https://jcenter.bintray.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
>       https://maven.google.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
>       https://maven.google.com//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
>       https://repo1.maven.org/maven2//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.pom
>       https://repo1.maven.org/maven2//android-beacon-library-2.13.1//android-beacon-library-2.13.1-.aar
>       file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/libs/android-beacon-library-2.13.1-.aar
>       file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/libs/android-beacon-library-2.13.1.aar
>   Required by:
>       project :app
>
> The very last path it tried seems close to correct because the altbeacon
> .aar file is located at
>
> file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/
> *src/main*/libs/android-beacon-library-2.13.1.aar
>
> Not
>
>
> file:/Users/jleveroni/Projects/joinify-dashboard/platforms/android/app/libs/android-beacon-library-2.13.1.aar
>
> Im not very experienced at all with gradle, is there a way to change the
> path that it searches for that .aar file in? Or perhaps my ionic build is
> misconfigured and the /libs folder is supposed to be directly under the
> app/ folder?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/petermetz/cordova-plugin-ibeacon/pull/364#issuecomment-382889517>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXqOL5LsTjfFdrXMQloh68R3wtYYVksks5tqQRbgaJpZM4TUXIx>
> .
>

petermetz commented on plugin.xml:
```diff
@@ -108,9 +108,7 @@
 				</intent-filter>
 			</receiver>
             <service android:enabled="true"
-                     android:exported="true"
```
@dopry Thank you for your contribution! Is there any specific reason for dropping the export of the service? Is this a necessary change for the version upgrade? 
If not needed for the upgrade, then please submit it as a separate PR (and then we can talk about the reasoning there)

dopry commented:
I was getting a conflict on the manifest merge for the export. I assumed it
was being exported upstream.

On Sun, Apr 22, 2018, 11:40 PM Peter Somogyvari <notifications@github.com>
wrote:

> *@petermetz* commented on this pull request.
> ------------------------------
>
> In plugin.xml
> <https://github.com/petermetz/cordova-plugin-ibeacon/pull/364#discussion_r183270931>
> :
>
> > @@ -108,9 +108,7 @@
>  				</intent-filter>
>  			</receiver>
>              <service android:enabled="true"
> -                     android:exported="true"
>
> @dopry <https://github.com/dopry> Thank you for your contribution! Is
> there any specific reason for dropping the export of the service? Is this a
> necessary change for the version upgrade?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/petermetz/cordova-plugin-ibeacon/pull/364#pullrequestreview-114236504>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXqOFeA9JcluWfh-MjvAIo6XGK7TiE5ks5trU0bgaJpZM4TUXIx>
> .
>

petermetz commented:
> I assumed it was being exported upstream.

It is. That's my point: it should stay that way unless it's necessary to change it. :-)

dopry commented:
I meant the upstream manifest has it marked as exported, so when the
service manifests are merged it ends up marked exported anyway and marking
it exported in the plugin manifest creates a conflict. I can undo it, but I
couldn't build with it there. Feel free to check out the branch and test
adding it back in.

On Mon, Apr 23, 2018, 12:51 AM Peter Somogyvari <notifications@github.com>
wrote:

> I assumed it was being exported upstream.
>
> It is. That's my point: it should stay that way unless it's necessary to
> change it. :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/petermetz/cordova-plugin-ibeacon/pull/364#issuecomment-383453618>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXqOLmwhH31DZnlYcf3wpJba3lnBwXjks5trV3ogaJpZM4TUXIx>
> .
>

dopry committed:
fix: add new permissions required for android 8

In testing errors were thrown without these permissions.

* BIND_JOB_SERVICE

```
Unhandled Promise rejection: No such service
ComponentInfo{com.sprygroup.automata/org.altbeacon.beacon.service.ScanJob} ; Zone: <root> ; Task: ; Value:
No such service ComponentInfo{com.sprygroup.automata/org.altbeacon.beacon.service.ScanJob}
```

* RECEIVE_BOOT_COMPLETED

```
Unhandled Promise rejection: Error: requested job be persisted without holding
RECEIVE_BOOT_COMPLETED permission. ; Zone: <root> ; Task: ; Value: Error: requested job be persisted without
holding RECEIVE_BOOT_COMPLETED permission.
```

dopry committed:
chore: remove configuration inherited altbeacon AndroidManifest.xml

Since AltBeacon is included as an dependency in grade through the `<framework />` tag
its AndroidManifest.xml is included during builds. The plugin no longer needs to reconstruct the
manifest simplifying manifest maintenance.

dopry commented:
@petermetz. I took a moment to do some additional testing and re-create the errors that led me to removeing the android:label and android:exported properties.

since I am including the class libraries as an AAR via the ```<framework />``` tag the AndroidManifest.xml inside the AAR is being merged in to the plugin.xml.  It creates a number of errors.

1. `Attribute service#org.altbeacon.beacon.service.BeaconService@label value=(iBeacon) from AndroidManifest.xml:25:97-120 is also present at [:android-beacon-library-2.13.1:] AndroidManifest.xml:34:13-35 value=(beacon). Suggestion: add 'tools:replace="android:label"' to <service> element at AndroidManifest.xml:25:9-27:19 to override.`

2. `Attribute service#org.altbeacon.beacon.service.BeaconService@exported value=(true) from AndroidManifest.xml:25:41-64 is also present at [:android-beacon-library-2.13.1:] AndroidManifest.xml:32:13-37 value=(false).  Suggestion: add 'tools:replace="android:exported"' to <service> element at AndroidManifest.xml:25:9-27:19 to override.`

Removing those attributes resolve the conflicts. Taking the idea that we can lean on the AltBeacon AndroidManifest.xml, I've removed all config that is already present upstream you don't have to maintain two sets of configuration.   

note, the PR is now based on #363, so merging this PR will bring in those commits.  The other can be merged first. I didn't want this PR to block the more basic fix since it introduces changes to the build.

jakeleveroni commented:
@petermetz @dopry does any more testing need to happen before this can be merged into the mainline? If so let me know I can help verify the changes work.

dopry commented:
@jakeleveroni,  I'm currently using it on projects. I'm satisfied with it. @petermetz, do you have any remaining reservations? Is there anything I need to address to get this merged?

petermetz commented:
@jakeleveroni @dopry Sorry, I'm very very busy, couldn't find any time for the project recently.
@dopry Would you like to have write access to the repo? I hate being a bottle-neck on the releases.

dopry commented:
If you're comfortable with that. I don't have the resources for testing on
iOS, but would be happy to help with the android side and to review code.

On Thu, Jun 14, 2018 at 6:19 PM, Peter Somogyvari <notifications@github.com>
wrote:

> @jakeleveroni <https://github.com/jakeleveroni> @dopry
> <https://github.com/dopry> Sorry, I'm very very busy, couldn't find any
> time for the project recently.
> @dopry <https://github.com/dopry> Would you like to have write access to
> the repo? I hate being a bottle-neck on the releases.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/petermetz/cordova-plugin-ibeacon/pull/364#issuecomment-397455573>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXqOM-by8BILiu03yrV668qgtbagngKks5t8uFZgaJpZM4TUXIx>
> .
>



-- 
Darrel O'Pry
The Spry Group, LLC.
http://www.spry-group.com
718-355-9767 x101

petermetz commented:
Okay, I'll give you the rights in a minute.
I was trying to revert this so that I can merge the other one first, but now I can't re-merge this one. Apologies. You may need to re-submit the changes in this for the GitHub UI to pick up the pace.

petermetz commented:
@dopry I've given you write access. It's fine if you only do Android changes, but at the same time feel free to merge in iOS related PRs as well, the community is usually very helpful in testing things out and coming back if there's new issues introduced with something. Thank you very much for any and all of your help on the project, I appreciate it.

dopry commented:
Cool... I'll try not to make a mess of everything. ;) Are there any
guidelines you'd like me to follow?

On Thu, Jun 14, 2018 at 6:26 PM, Peter Somogyvari <notifications@github.com>
wrote:

> @dopry <https://github.com/dopry> I've given you write access. It's fine
> if you only do Android changes, but at the same time feel free to merge in
> iOS related PRs as well, the community is usually very helpful in testing
> things out and coming back if there's new issues introduced with something.
> Thank you very much for any and all of your help on the project, I
> appreciate it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/petermetz/cordova-plugin-ibeacon/pull/364#issuecomment-397457045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXqOEK-XfA6wyBx0EnNRXAC_oeYN-jnks5t8uMEgaJpZM4TUXIx>
> .
>



-- 
Darrel O'Pry
The Spry Group, LLC.
http://www.spry-group.com
718-355-9767 x101

petermetz commented:
Awesome. :-)
Nothing too specific regarding the guidelines that aren't in most best practice collections already.
If you are looking at someone else's PR always make sure to know and understand each line of change. If something is not obvious to you, it probably won't be to anybody else either, so in that case it has to be explicitly documented in the commit message.

I always ask people to avoid committing white space/newline changes that clutter the diffs.

The versioning is semver, breaking changes can happen but we'll need to bump from 3.x to 4.x.

I'll try to take a look at npm, see if I could grant you access to push releases there as well.

dopry commented:
alright.  oh if you want me to clean up the git history. I can reset 3.x to before the merge and revert and re-apply the branches from my PRs cleanly.  It'll only take me a few seconds.

petermetz commented:
Well, normally I wouldn't want to bother you with something like that, but if it really is just a few seconds then it would be very much appreciated. ;-)

dopry commented:
Done. I'm used to working with rebase, reset, and squash heavy workflows, I've developed a good hand for re-writing history.

dopry commented:
remerged as https://github.com/petermetz/cordova-plugin-ibeacon/commit/1a637b8d2dd028cf56963fcc678a352b5788f0c9
